File: integrations/title_cleaning.py
import re


LEADING_SYSTEM_ALIASES = [
    "Nintendo Switch 2",
    "Nintendo Switch",
    "Nintendo 3DS",
    "Nintendo DS",
    "Nintendo Wii U",
    "Nintendo Wii",
    "Playstation 5",
    "PlayStation 5",
    "Playstation 4",
    "PlayStation 4",
    "Playstation 3",
    "PlayStation 3",
    "Playstation 2",
    "PlayStation 2",
    "Xbox Series X",
    "Xbox Series S",
    "Xbox One",
    "Xbox 360",
    "Gamecube",
    "Nintendo 64",
    "Super Nintendo",
    "PS Vita",
    "Switch 2",
    "Switch",
    "PS5",
    "PS4",
    "PS3",
    "PS2",
    "PSP",
    "NES",
    "DS",
    "PC",
    "Wii U",
    "Wii",
    "Xbox",
]

NOISE_PHRASES = [
    "new:",
    "brand new sealed",
    "brand new / sealed",
    "brand new/sealed",
    "brand new factory sealed read",
    "brand new factory sealed",
    "factory sealed",
    "free shipping",
    "fast shipping",
    "ships fast",
    "please read",
    "read description",
    "new sealed",
]

NOISE_WORDS = {
    "brand",
    "factory",
    "fast",
    "free",
    "new",
    "read",
    "sealed",
    "shipping",
    "ships",
}


def clean_marketplace_title_for_search(title: str | None) -> str:
    """Clean an eBay/marketplace title for Amazon catalog search.

    This keeps useful title and platform terms, removes common condition/shipping
    noise, drops release years inside system parentheticals, and moves leading
    system terms to the end of the query.
    """
    if not title:
        return ""

    title_with_clean_parentheticals = remove_parenthetical_release_years(title)
    title_without_noise = trim_noise_words(
        remove_noise_phrases(title_with_clean_parentheticals)
    )
    cleaned_title, leading_systems = move_leading_system_terms(title_without_noise)
    search_term = cleanup_search_text(
        " ".join([cleaned_title, " ".join(leading_systems)])
    )

    return search_term


def remove_noise_phrases(value: str) -> str:
    current_value = value

    for phrase in NOISE_PHRASES:
        current_value = re.sub(
            rf"(^|\b){re.escape(phrase)}(?=\s|\b|$)",
            " ",
            current_value,
            flags=re.IGNORECASE,
        )

    return current_value


def remove_parenthetical_release_years(value: str) -> str:
    def replace_match(match):
        contents = match.group(1)
        cleaned_contents = cleanup_search_text(
            re.sub(r"\b(?:19|20)\d{2}\b", " ", contents)
        )

        return f" {cleaned_contents} " if cleaned_contents else " "

    return re.sub(r"\(([^)]*)\)", replace_match, value)


def trim_noise_words(value: str) -> str:
    words = cleanup_search_text(value).split(" ")

    while words and words[0].lower() in NOISE_WORDS:
        words.pop(0)

    while words and words[-1].lower() in NOISE_WORDS:
        words.pop()

    return " ".join(words)


def move_leading_system_terms(value: str) -> tuple[str, list[str]]:
    leading_systems = []
    remaining_title = cleanup_search_text(value)
    matched = True

    while matched:
        matched = False

        for alias in LEADING_SYSTEM_ALIASES:
            match = re.match(
                rf"^{re.escape(alias)}(?=\s|[-:/|]|$)",
                remaining_title,
                flags=re.IGNORECASE,
            )

            if not match:
                continue

            leading_systems.append(match.group(0))
            remaining_title = cleanup_search_text(
                re.sub(r"^[-:/|\s]+", "", remaining_title[match.end():])
            )
            matched = True
            break

    return remaining_title, leading_systems


def cleanup_search_text(value: str) -> str:
    text = re.sub(r"[()[\]{}]", " ", value or "")
    text = re.sub(r"[.,!?]+", " ", text)
    text = re.sub(r"\s*[/|]\s*", " ", text)
    text = re.sub(r"\s*[-:]\s*", " ", text)
    text = re.sub(r"\s+", " ", text)
    return text.strip()

File: web/app/purchases/titleCleaning.ts
const LEADING_SYSTEM_ALIASES = [
  "Nintendo Switch 2",
  "Nintendo Switch",
  "Nintendo 3DS",
  "Nintendo DS",
  "Nintendo Wii U",
  "Nintendo Wii",
  "Playstation 5",
  "PlayStation 5",
  "Playstation 4",
  "PlayStation 4",
  "Playstation 3",
  "PlayStation 3",
  "Playstation 2",
  "PlayStation 2",
  "Xbox Series X",
  "Xbox Series S",
  "Xbox One",
  "Xbox 360",
  "Gamecube",
  "Nintendo 64",
  "Super Nintendo",
  "PS Vita",
  "Switch 2",
  "Switch",
  "PS5",
  "PS4",
  "PS3",
  "PS2",
  "PSP",
  "NES",
  "DS",
  "PC",
  "Wii U",
  "Wii",
  "Xbox",
];

const NOISE_PHRASES = [
  "brand new factory sealed read",
  "brand new factory sealed",
  "complete with manual tested works",
  "brand new and sealed",
  "multiplayer music game",
  "brand new / sealed",
  "brand new/sealed",
  "brand new sealed",
  "complete video game",
  "free us shipping",
  "complete with manual",
  "action & adventure",
  "new and sealed",
  "new & sealed",
  "read description",
  "factory sealed",
  "fast shipping",
  "fast slipping",
  "still sealed",
  "sealed new",
  "new unopened",
  "1st printing",
  "first printing",
  "tested works",
  "video game",
  "electronic arts",
  "warner bros multiplayer",
  "free shipping",
  "please read",
  "new sealed",
  "ships fast",
  "dvd-rom",
  "dvd rom",
  "2k games",
  "firefly studios",
  "ea / dice",
  "ntsc-u/c",
  "ntsc u c",
  "action adventure",
  "us version",
  "us edition",
  "game esrb",
  "brand new",
  "unopened",
  "ea dice",
  "read a10",
  "slipping",
  "new:",
  "dvd",
  "esrb",
  "promo",
];

const NOISE_WORDS = new Set([
  "brand",
  "factory",
  "fast",
  "free",
  "read",
  "sealed",
  "shipping",
  "ships",
  "microsoft",
  "sony",
]);

const TRIM_NOISE_WORDS = NOISE_WORDS;

export function cleanMarketplaceTitleForSearch(title: string) {
  if (/\bwii\s+play\b/i.test(title)) return "Wii Play";

  const titleWithCleanParentheticals = removeParentheticalReleaseYears(title);
  const titleWithoutNoise = removeNoiseWords(
    trimNoiseWords(removeNoisePhrases(titleWithCleanParentheticals))
  );
  const { title: cleanedTitle, leadingSystems } =
    moveLeadingSystemTerms(titleWithoutNoise);
  const searchTerm = cleanupSearchText(
    [cleanedTitle, leadingSystems.join(" ")].join(" ")
  );

  return searchTerm;
}

function removeNoisePhrases(value: string) {
  return NOISE_PHRASES.reduce(
    (currentValue, phrase) =>
      currentValue.replace(
        new RegExp(`(^|\\b)${escapeRegExp(phrase)}(?=\\s|\\b|$)`, "gi"),
        " "
      ),
    value
  );
}

function removeParentheticalReleaseYears(value: string) {
  return value.replace(/\(([^)]*)\)/g, (_match, contents: string) => {
    const cleanedContents = cleanupSearchText(
      contents.replace(/\b(?:19|20)\d{2}\b/g, " ")
    );

    return cleanedContents ? ` ${cleanedContents} ` : " ";
  });
}

function trimNoiseWords(value: string) {
  const words = cleanupSearchText(value).split(" ");

  while (words.length > 0 && TRIM_NOISE_WORDS.has(words[0].toLowerCase())) {
    words.shift();
  }

  while (words.length > 0 && TRIM_NOISE_WORDS.has(words[words.length - 1].toLowerCase())) {
    words.pop();
  }

  return words.join(" ");
}

function removeNoiseWords(value: string) {
  return cleanupSearchText(value)
    .split(" ")
    .filter((word) => !NOISE_WORDS.has(word.toLowerCase()))
    .join(" ");
}

function moveLeadingSystemTerms(value: string) {
  const leadingSystems: string[] = [];
  let remainingTitle = cleanupSearchText(value);
  let matched = true;

  while (matched) {
    matched = false;

    for (const alias of LEADING_SYSTEM_ALIASES) {
      const pattern = new RegExp(`^${escapeRegExp(alias)}(?=\\s|[-:/|]|$)`, "i");
      const match = remainingTitle.match(pattern);

      if (!match) continue;

      leadingSystems.push(match[0]);
      remainingTitle = cleanupSearchText(
        remainingTitle.slice(match[0].length).replace(/^[-:/|\s]+/, "")
      );
      matched = true;
      break;
    }
  }

  return {
    title: remainingTitle,
    leadingSystems,
  };
}

function cleanupSearchText(value: string) {
  return value
    .replace(/[()[\]{}"]/g, " ")
    .replace(/[.,!?*]+/g, " ")
    .replace(/\s*[/|]\s*/g, " ")
    .replace(/\s*[-:\u2013\u2014\u2022]\s*/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function escapeRegExp(value: string) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
